File: bridges/nodejs/src/sdk/tools/faster_whisper-tool.ts
import { Tool } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'

const MODEL_NAME = 'faster-whisper-large-v3'

export default class FasterWhisperTool extends Tool {
  private static readonly TOOLKIT = 'music_audio'
  private readonly config: ReturnType<typeof ToolkitConfig.load>

  constructor() {
    super()
    // Load configuration from central toolkits directory
    this.config = ToolkitConfig.load(FasterWhisperTool.TOOLKIT, this.toolName)
  }

  get toolName(): string {
    // Use the actual config name for toolkit lookup
    return 'faster_whisper'
  }

  get toolkit(): string {
    return FasterWhisperTool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Transcribe audio to a file using faster-whisper
   * @param inputPath The file path of the audio to be transcribed
   * @param outputPath The desired file path for the transcription output
   * @param device Device to use for processing (cpu, cuda, auto)
   * @param cpuThreads Number of CPU threads to use
   * @param downloadRoot Root directory for model downloads
   * @param localFilesOnly Whether to use only local files
   * @returns A promise that resolves with the path to the transcription file
   */
  async transcribeToFile(
    inputPath: string,
    outputPath: string,
    device = 'auto',
    cpuThreads?: number,
    downloadRoot?: string,
    localFilesOnly = false
  ): Promise<string> {
    try {
      // Get model path using the generic resource system
      const modelPath = await this.getResourcePath(MODEL_NAME)

      const args = [
        '--function',
        'transcribe_to_file',
        '--input',
        inputPath,
        '--output',
        outputPath,
        '--model_size_or_path',
        modelPath,
        '--device',
        device
      ]

      if (cpuThreads) {
        args.push('--cpu_threads', cpuThreads.toString())
      }

      if (downloadRoot) {
        args.push('--download_root', downloadRoot)
      }

      if (localFilesOnly) {
        args.push('--local_files_only')
      }

      await this.executeCommand({
        binaryName: 'faster_whisper',
        args,
        options: { sync: true }
      })

      return outputPath
    } catch (error: unknown) {
      throw new Error(`Audio transcription failed: ${(error as Error).message}`)
    }
  }
}

File: bridges/python/src/sdk/tools/faster_whisper_tool.py
from typing import Optional

from ..base_tool import BaseTool, ExecuteCommandOptions
from ..toolkit_config import ToolkitConfig

MODEL_NAME = 'faster-whisper-large-v3'

class FasterWhisperTool(BaseTool):
    TOOLKIT = 'music_audio'

    def __init__(self):
        super().__init__()
        # Load configuration from central toolkits directory
        self.config = ToolkitConfig.load(self.TOOLKIT, self.tool_name)

    @property
    def tool_name(self) -> str:
        # Use the actual config name for toolkit lookup
        return 'faster_whisper'

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config['description']

    def transcribe_to_file(
        self,
        input_path: str,
        output_path: str,
        device: str = 'auto',
        cpu_threads: Optional[int] = None,
        download_root: Optional[str] = None,
        local_files_only: bool = False
    ) -> str:
        """
        Transcribe audio to a file using faster-whisper

        Args:
            input_path: The file path of the audio to be transcribed
            output_path: The desired file path for the transcription output
            device: Device to use for processing (cpu, cuda, auto)
            cpu_threads: Number of CPU threads to use
            download_root: Root directory for model downloads
            local_files_only: Whether to use only local files

        Returns:
            The path to the transcription file
        """
        try:
            # Get model path using the generic resource system
            model_path = self.get_resource_path(MODEL_NAME)

            args = [
                '--function', 'transcribe_to_file',
                '--input', input_path,
                '--output', output_path,
                '--model_size_or_path', model_path,
                '--device', device
            ]

            if cpu_threads:
                args.extend(['--cpu_threads', str(cpu_threads)])

            if download_root:
                args.extend(['--download_root', download_root])

            if local_files_only:
                args.append('--local_files_only')

            self.execute_command(ExecuteCommandOptions(
                binary_name='faster_whisper',
                args=args,
                options={'sync': True}
            ))

            return output_path
        except Exception as e:
            raise Exception(f"Audio transcription failed: {str(e)}")
